File: src/main/java/jkmau5/alternativeenergy/gui/container/AltEngContainer.java
package jkmau5.alternativeenergy.gui.container;

import com.google.common.collect.Lists;
import cpw.mods.fml.common.network.PacketDispatcher;
import cpw.mods.fml.common.network.Player;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import jkmau5.alternativeenergy.gui.container.slot.AltEngSlot;
import jkmau5.alternativeenergy.gui.element.Element;
import jkmau5.alternativeenergy.inventory.InventoryUtils;
import jkmau5.alternativeenergy.network.PacketElementUpdate;
import lombok.Getter;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.ICrafting;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;

import java.io.DataInput;
import java.io.IOException;
import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public abstract class AltEngContainer extends Container {

    private final IInventory inventory;
    @Getter private List<Element> elements = Lists.newArrayList();

    public AltEngContainer(IInventory inventory) {
        this.inventory = inventory;
    }

    public AltEngContainer() {
        this.inventory = null;
    }

    public void addSlot(Slot slot) {
        this.addSlotToContainer(slot);
    }

    public void addElement(Element element) {
        element.setContainer(this);
        this.elements.add(element);
    }

    @Override
    public void addCraftingToCrafters(ICrafting crafter) {
        super.addCraftingToCrafters(crafter);
        for(Element element : this.elements) {
            element.initElement(crafter);
        }
    }

    @Override
    public void detectAndSendChanges() {
        super.detectAndSendChanges();
        for(Element element : this.elements) {
            for(ICrafting crafter : (List<ICrafting>) this.crafters) {
                element.updateElement(crafter);
            }
        }
    }

    public void sendElementDataToClient(Element element, ICrafting crafter, byte[] data) {
        PacketDispatcher.sendPacketToPlayer(new PacketElementUpdate(this.windowId, this.elements.indexOf(element), data).getPacket(), (Player) crafter);
    }

    public void handleElementDataClient(int widgetId, DataInput data) throws IOException {
        this.elements.get(widgetId).handleClientPacketData(data);
    }

    @SideOnly(Side.CLIENT)
    public void updateString(byte id, String data) {
    }

    public boolean canInteractWith(EntityPlayer entityplayer) {
        if(this.inventory == null) {
            return true;
        }
        return this.inventory.isUseableByPlayer(entityplayer);
    }

    protected void addPlayerInventory(InventoryPlayer inventoryPlayer) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 9; j++) {
                this.addSlot(new Slot(inventoryPlayer, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
            }
        }
        for (int i = 0; i < 9; i++) {
            this.addSlot(new Slot(inventoryPlayer, i, 8 + i * 18, 142));
        }
    }

    public ItemStack slotClick(int slotNum, int mouseButton, int modifier, EntityPlayer player) {
        Slot slot = slotNum < 0 ? null : (Slot)this.inventorySlots.get(slotNum);
        if(slot instanceof AltEngSlot && ((AltEngSlot) slot).isFakeSlot()) {
            return this.fakeSlotClick((AltEngSlot) slot, mouseButton, modifier, player);
        }
        return super.slotClick(slotNum, mouseButton, modifier, player);
    }

    private ItemStack fakeSlotClick(AltEngSlot slot, int mouseButton, int modifier, EntityPlayer player) {
        ItemStack stack = null;

        if(mouseButton == 2) {
            if(slot.canAdjustFakeSlot()) {
                slot.putStack(null);
            }
        } else if((mouseButton == 0) || (mouseButton == 1)) {
            InventoryPlayer playerInv = player.inventory;
            slot.onSlotChanged();
            ItemStack stackSlot = slot.getStack();
            ItemStack stackHeld = playerInv.getItemStack();

            if (stackSlot != null) {
                stack = stackSlot.copy();
            }
            if (stackSlot == null) {
                if (stackHeld != null && slot.isItemValid(stackHeld)) {
                    this.fillFakeSlot(slot, stackHeld, mouseButton, modifier);
                }
            } else if(stackHeld == null) {
                adjustFakeSlot(slot, mouseButton, modifier);
                slot.onPickupFromSlot(player, playerInv.getItemStack());
            } else if(slot.isItemValid(stackHeld)) {
                if(InventoryUtils.areItemsEqual(stackSlot, stackHeld)) {
                    this.adjustFakeSlot(slot, mouseButton, modifier);
                } else {
                    this.fillFakeSlot(slot, stackHeld, mouseButton, modifier);
                }
            }
        }
        return stack;
    }

    protected void adjustFakeSlot(AltEngSlot slot, int mouseButton, int modifier) {
        if(!slot.canAdjustFakeSlot()) {
            return;
        }
        ItemStack stackSlot = slot.getStack();
        int stackSize;
        if (modifier == 1) {
            stackSize = mouseButton == 0 ? (stackSlot.stackSize + 1) / 2 : stackSlot.stackSize * 2;
        } else {
            stackSize = mouseButton == 0 ? stackSlot.stackSize - 1 : stackSlot.stackSize + 1;
        }
        if(stackSize > slot.getSlotStackLimit()) {
            stackSize = slot.getSlotStackLimit();
        }
        stackSlot.stackSize = stackSize;
        if (stackSlot.stackSize <= 0) {
            slot.putStack(null);
        }
    }

    protected void fillFakeSlot(AltEngSlot slot, ItemStack stackHeld, int mouseButton, int modifier) {
        if(!slot.canAdjustFakeSlot()) {
            return;
        }
        int stackSize = mouseButton == 0 ? stackHeld.stackSize : 1;
        if (stackSize > slot.getSlotStackLimit()) {
            stackSize = slot.getSlotStackLimit();
        }
        ItemStack phantomStack = stackHeld.copy();
        phantomStack.stackSize = stackSize;
        slot.putStack(phantomStack);
    }

    protected boolean shiftItemStack(ItemStack stackToShift, int start, int end) {
        boolean changed = false;
        if (stackToShift.isStackable()) {
            for (int slotIndex = start; stackToShift.stackSize > 0 && slotIndex < end; slotIndex++) {
                Slot slot = (Slot) this.inventorySlots.get(slotIndex);
                ItemStack stackInSlot = slot.getStack();
                if (stackInSlot != null && InventoryUtils.areItemsEqual(stackInSlot, stackToShift)) {
                    int resultingStackSize = stackInSlot.stackSize + stackToShift.stackSize;
                    int max = Math.min(stackToShift.getMaxStackSize(), slot.getSlotStackLimit());
                    if (resultingStackSize <= max) {
                        stackToShift.stackSize = 0;
                        stackInSlot.stackSize = resultingStackSize;
                        slot.onSlotChanged();
                        changed = true;
                    } else if(stackInSlot.stackSize < max) {
                        stackToShift.stackSize -= max - stackInSlot.stackSize;
                        stackInSlot.stackSize = max;
                        slot.onSlotChanged();
                        changed = true;
                    }
                }
            }
        }
        if(stackToShift.stackSize > 0) {
            for (int slotIndex = start; stackToShift.stackSize > 0 && slotIndex < end; slotIndex++) {
                Slot slot = (Slot) this.inventorySlots.get(slotIndex);
                ItemStack stackInSlot = slot.getStack();
                if(stackInSlot == null) {
                    int max = Math.min(stackToShift.getMaxStackSize(), slot.getSlotStackLimit());
                    stackInSlot = stackToShift.copy();
                    stackInSlot.stackSize = Math.min(stackToShift.stackSize, max);
                    stackToShift.stackSize -= stackInSlot.stackSize;
                    slot.putStack(stackInSlot);
                    slot.onSlotChanged();
                    changed = true;
                }
            }
        }
        return changed;
    }

    private boolean tryShiftItem(ItemStack stackToShift, int numSlots) {
        for (int machineIndex = 0; machineIndex < numSlots - 36; machineIndex++) {
            Slot slot = (Slot) this.inventorySlots.get(machineIndex);
            if (!(slot instanceof AltEngSlot) || ((AltEngSlot) slot).canShiftClick()) {
                if(!(slot instanceof AltEngSlot) || ((AltEngSlot) slot).isFakeSlot()) {
                    if (slot.isItemValid(stackToShift)) {
                        if (shiftItemStack(stackToShift, machineIndex, machineIndex + 1)) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    public ItemStack transferStackInSlot(EntityPlayer player, int slotIndex) {
        ItemStack originalStack = null;
        Slot slot = (Slot) this.inventorySlots.get(slotIndex);
        int numSlots = this.inventorySlots.size();
        if(slot != null && slot.getHasStack()) {
            ItemStack stackInSlot = slot.getStack();
            originalStack = stackInSlot.copy();
            if (slotIndex < numSlots - 36 || !tryShiftItem(stackInSlot, numSlots)) {
                if(slotIndex >= numSlots - 36 && slotIndex < numSlots - 9) {
                    if(!shiftItemStack(stackInSlot, numSlots - 9, numSlots)) {
                        return null;
                    }
                } else if(slotIndex >= numSlots - 9 && slotIndex < numSlots) {
                    if(!shiftItemStack(stackInSlot, numSlots - 36, numSlots - 9)) {
                        return null;
                    }
                } else if(!shiftItemStack(stackInSlot, numSlots - 36, numSlots)) {
                    return null;
                }
            }
            slot.onSlotChange(stackInSlot, originalStack);
            if (stackInSlot.stackSize <= 0) {
                slot.putStack(null);
            } else {
                slot.onSlotChanged();
            }
            if (stackInSlot.stackSize == originalStack.stackSize) {
                return null;
            }
            slot.onPickupFromSlot(player, stackInSlot);
        }
        return originalStack;
    }
}

File: src/main/scala/alternativemods/alteng/gui/containers/AltEngContainer.scala
package alternativemods.alteng.gui.containers

import net.minecraft.inventory.{ICrafting, Slot, IInventory, Container}
import net.minecraft.entity.player.{InventoryPlayer, EntityPlayer}
import scala.collection.mutable
import scala.collection.JavaConversions._
import alternativemods.alteng.gui.element.Element
import io.netty.buffer.ByteBuf
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.item.ItemStack
import alternativemods.alteng.gui.containers.slot.AltEngSlot
import alternativemods.alteng.util.InventoryUtils

/**
 * No description given
 *
 * @author jk-5
 */
abstract class AltEngContainer(private final val inventory: Option[IInventory] = None) extends Container {

  val elements = mutable.ArrayBuffer[Element]()

  final def addSlot(slot: Slot) = this.addSlotToContainer(slot)
  final def addElement(element: Element){
    element.container = this
    this.elements += element
  }

  override def addCraftingToCrafters(crafter: ICrafting){
    super.addCraftingToCrafters(crafter)
    this.elements.foreach(_.initElement(crafter))
  }

  override def detectAndSendChanges(){
    super.detectAndSendChanges()
    this.elements.foreach(e => this.crafters.map(_.asInstanceOf[ICrafting]).foreach(c => e.updateElement(c)))
  }

  def sendElementDataToClient(element: Element, crafter: ICrafting, data: ByteBuf){
    //TODO
  }

  def handleElementData(id: Int, data: ByteBuf){
    this.elements(id).handleClientPacket(data)
  }

  @SideOnly(Side.CLIENT) def updateString(id: Byte, data: String){}

  override def canInteractWith(player: EntityPlayer) = if(this.inventory.isEmpty) true else this.inventory.get.isUseableByPlayer(player)

  protected def addPlayerInventory(inventory: InventoryPlayer){
    0 to 2 foreach(y => 0 to 8 foreach(x =>
      this.addSlot(new Slot(inventory, x + y * 9 + 9, 8 + x * 18, 84 + y * 18))
    ))
    0 to 8 foreach(i => this.addSlot(new Slot(inventory, i, 8 + i * 18, 142)))
  }

  override def slotClick(slotId: Int, button: Int, modifier: Int, player: EntityPlayer): ItemStack = {
    val slot = if(slotId < 0) null else this.inventorySlots.get(slotId).asInstanceOf[Slot]
    slot match {
      case s: AltEngSlot => if(s.isFakeSlot) this.fakeSlotClick(s, button, modifier, player)
        else super.slotClick(slotId, button, modifier, player)
      case _ => super.slotClick(slotId, button, modifier, player)
    }
  }

  private def fakeSlotClick(slot: AltEngSlot, button: Int, modifier: Int, player: EntityPlayer): ItemStack = {
    var stack: ItemStack = null

    if(button == 2){
      if(slot.canAdjustFakeSlot) slot.putStack(null)
    }else if(button == 0 || button == 1){
      val playerInv = player.inventory
      slot.onSlotChanged()
      val stackSlot = slot.getStack
      val stackHeld = playerInv.getItemStack

      if(stackSlot != null) stack = stackSlot.copy()
      if(stackSlot == null){
        if(stackHeld != null && slot.isItemValid(stackHeld)){
          this.fillFakeSlot(slot, stackHeld, button, modifier)
        }
      }else if(stackHeld == null){
        adjustFakeSlot(slot, button, modifier)
        slot.onPickupFromSlot(player, playerInv.getItemStack)
      }else if(slot.isItemValid(stackHeld)){
        if(InventoryUtils.areItemsEqual(stackSlot, stackHeld)) {
          this.adjustFakeSlot(slot, button, modifier)
        }else{
          this.fillFakeSlot(slot, stackHeld, button, modifier)
        }
      }
    }
    stack
  }

  protected def adjustFakeSlot(slot: AltEngSlot, button: Int, modifier: Int){
    if(!slot.canAdjustFakeSlot) return
    val stackSlot = slot.getStack
    var stackSize = 0
    if(modifier == 1){
      stackSize = if(button == 0) (stackSlot.stackSize + 1) / 2 else stackSlot.stackSize * 2
    }else{
      stackSize = if(button == 0) stackSlot.stackSize - 1 else stackSlot.stackSize + 1
    }
    stackSize = Math.min(stackSize, slot.getSlotStackLimit)
    stackSlot.stackSize = stackSize
    if(stackSlot.stackSize <= 0){
      slot.putStack(null)
    }
  }

  protected def fillFakeSlot(slot: AltEngSlot, held: ItemStack, button: Int, modifier: Int){
    if(!slot.canAdjustFakeSlot) return
    var stackSize = if(button == 0) held.stackSize else 1
    stackSize = Math.min(stackSize, slot.getSlotStackLimit)
    val phantomStack = held.copy()
    phantomStack.stackSize = stackSize
    slot.putStack(phantomStack)
  }

  def shiftItemStack(stackToShift: ItemStack, start: Int, end: Int): Boolean = {
    var changed = false
    if(stackToShift.isStackable){
      var slotIndex = start
      while(stackToShift.stackSize > 0 && slotIndex < end){
        val slot = this.inventorySlots.get(slotIndex).asInstanceOf[Slot]
        val stackInSlot = slot.getStack
        if(stackInSlot != null && InventoryUtils.areItemsEqual(stackInSlot, stackToShift)){
          val resultingStackSize = stackInSlot.stackSize + stackToShift.stackSize
          val max = Math.min(stackToShift.getMaxStackSize, slot.getSlotStackLimit)
          if(resultingStackSize <= max){
            stackToShift.stackSize = 0
            stackInSlot.stackSize = resultingStackSize
            slot.onSlotChanged()
            changed = true
          }else if(stackInSlot.stackSize < max){
            stackToShift.stackSize -= max - stackInSlot.stackSize
            stackInSlot.stackSize = max
            slot.onSlotChanged()
            changed = true
          }
        }
        slotIndex += 1
      }
    }
    if(stackToShift.stackSize > 0) {
      var slotIndex = start
      while(stackToShift.stackSize > 0 && slotIndex < end){
        val slot = this.inventorySlots.get(slotIndex).asInstanceOf[Slot]
        var stackInSlot = slot.getStack
        if(stackInSlot == null) {
          val max = Math.min(stackToShift.getMaxStackSize, slot.getSlotStackLimit)
          stackInSlot = stackToShift.copy()
          stackInSlot.stackSize = Math.min(stackToShift.stackSize, max)
          stackToShift.stackSize -= stackInSlot.stackSize
          slot.putStack(stackInSlot)
          slot.onSlotChanged()
          changed = true
        }
        slotIndex += 1
      }
    }
    changed
  }

  private def tryShiftItem(stackToShift: ItemStack, numSlots: Int): Boolean = {
    for(machineIndex <- 0 until numSlots - 36){
      val slot = this.inventorySlots.get(machineIndex).asInstanceOf[Slot]
      if(!slot.isInstanceOf[AltEngSlot] || slot.asInstanceOf[AltEngSlot].canShiftClick){
        if(!slot.isInstanceOf[AltEngSlot] || slot.asInstanceOf[AltEngSlot].isFakeSlot){
          if(slot.isItemValid(stackToShift)){
            if(shiftItemStack(stackToShift, machineIndex, machineIndex + 1)){
              return true
            }
          }
        }
      }
    }
    false
  }

  override def transferStackInSlot(player: EntityPlayer, index: Int): ItemStack = {
    var originalStack: ItemStack = null
    val slot = this.inventorySlots.get(index).asInstanceOf[Slot]
    val numSlots = this.inventorySlots.size()
    if(slot != null && slot.getHasStack){
      val stackInSlot = slot.getStack
      originalStack = stackInSlot.copy()
      if(index < numSlots - 36 || !tryShiftItem(stackInSlot, numSlots)){
        if(index >= numSlots - 36 && index < numSlots - 9){
          if(!shiftItemStack(stackInSlot, numSlots - 9, numSlots)){
            return null
          }
        }else if(index >= numSlots - 9 && index < numSlots){
          if(!shiftItemStack(stackInSlot, numSlots - 36, numSlots - 9)){
            return null
          }
        }else if(!shiftItemStack(stackInSlot, numSlots - 36, numSlots)){
          return null
        }
      }
      slot.onSlotChange(stackInSlot, originalStack)
      if(stackInSlot.stackSize <= 0){
        slot.putStack(null)
      }else{
        slot.onSlotChanged()
      }
      if(stackInSlot.stackSize == originalStack.stackSize){
        return null
      }
      slot.onPickupFromSlot(player, stackInSlot)
    }
    originalStack
  }
}
